Refactor Structure

THEtheChad opened the pull request:
I propose reorganizing the project structure a bit. I think it makes sense to move the fixture generators into the `fixtures` folder and renaming them from `defaultGenerators` to `fixtureGenerators`. I also renamed `main.ts` to `browser.ts` because I felt it could be misleading as to its purpose (it's definitely not main to anything).

timdeschryver commented on src/core/core.test.ts:
```diff
@@ -15,7 +15,7 @@ describe('core', () => {
 
 	test('creates a fixture', () => {
-		const core = new Core().register(defaultGenerators);
+		const core = new Core().register(fixtureGenerators);
```
I like the restructure, but here I am again bringing up the name `Core` 😅
I wonder if we can rename it to something else.
Core makes sense for the internals, but I don't know if it's a good name to export.

`Core` doesn't make the intent clear if a user would import it (and maybe there are multiple libraries that export a `Core`).

The name I bring up again is `Fixture`.
And I would like to hear your thoughts on the following example(s):

```ts
// manual
const fixture = new Fixture().register(fixtureGenerators);
const value = fixture.generate(zodSchema) 

// predefined
// this is different than the current API because 
// createFixture returns the preconfigured fixture including the default generators
// to generate a value, generate needs to be invoked
const fixture = createFixture() // implementation would be new Fixture().register(defaultGenerators);
const value = fixture.generate(zodSchema) // or fixture.create

// predefined + extended
const fixture = createFixture() 
fixture.extend(customGenerator)
const value = fixture.generate(zodSchema)  // or fixture.create


// serialize (also preconfigured)
const fixture = createSerializableFixture() 
const value = fixture.generate(zodSchema)  // or fixture.create
```

This would also make it possible to reuse a fixture across multiple tests with the same seed, which isn't possible with `createFixture` without updating all references.


I promise that it's the last time that I bring this up 😅

THEtheChad replied:
No no, keep bringing things up! These are good thought exercises and valid points. You're definitely on to something with the ability to set a global seed value and core IS very generic and doesn't describe what the class is actually doing. Let me contemplate these issues and explore alternatives.

THEtheChad replied:
I've been thinking about what `Core` is and what `Core` does. It's essentially a "transpiler" that takes a Zod schema and transforms it into another object with a similar structure.

```ts
// manual
const fixture = new Transpiler({ seed: 1234 }).extend(fixtureGenerators);
const value = fixture.from(zodSchema) // same as fixture.generate

// we can now extend the Transpiler class
export class Fixture extends Transpiler {
	generators = fixtureGenerators;

	// explicitly define the return type
	generate<TSchema extends z.ZodTypeAny>(
		schema: TSchema,
		context?: Context
	): z.infer<TSchema> {
		return super.generate(schema, context);
	}
}

// predefined
// this is different than the current API because 
const fixture = new Fixture({ seed: 1234 })
const value = fixture.from(zodSchema) // same as fixture.generate

// predefined + extended
const fixture =  new Fixture({ seed: 1234 })
fixture.extend(customGenerator)
const value = fixture.from(zodSchema)  // same as fixture.generate

// serializer (also preconfigured by extending Transpiler)
const serializer = new Serializer({ seed: 1234 })
const value = serializer.from(zodSchema)  // same as serializer.generate
```

THEtheChad replied:
Or possibly Translate or Translator?

timdeschryver replied:
I like `Transformer` the most

THEtheChad commented:
Lots of files touched because of the name change but the major highlights are:
- `Core` is now `Transformer` in call cases except the generator calls. Should we keep it this way or make a change there, too?
- `main.ts` is now `browser.ts` to highlight the fact that it's just browser based "testing". Maybe it should be moved into the `test` folder?
- `index.ts` changed to `export.ts` to clarify what's actually being published as the public API. This is a single commit so it's easy to revert if you don't like it.
- Folder structure has changed. `generators` have been colocated alongside the transformers they're utilized with.

I still don't like the syntax for subclassing a `Transformer` but it's a low priority to explore since 99% of people won't be subclassing.

timdeschryver commented:
Nice! A lot of changes but now it the time to do it :)

- I think we should also rename `core` in the generator 
- I don't have a strong opinion about it, but I think it's fine where it's now
- In the "Angular world" we call it a `public-api.ts` file, but I don't have a strong opinion about this either. The only concern I have is that `export` might be confusing because it's a reserved keyword.
- I like it
- `.from` starts growing on me, it's easy to read 👍
- I'll thinker on the subclassing later, but I agree that it's a low prio

I'll update the readme when this PR is merged :)

THEtheChad commented:
> I think we should also rename core in the generator

All references to the `core` instance have been changed to `transform` and I kind of like it 🥳

> I don't have a strong opinion about it, but I think it's fine where it's now

I moved `browser.ts`  under the `test` folder and it feels nice to have it out of the core codebase

> In the "Angular world" we call it a public-api.ts file, but I don't have a strong opinion about this either. The only concern I have is that export might be confusing because it's a reserved keyword.

I renamed it to `public.ts` because I don't like the hyphen and I think it's self explanatory 🙃

> I like it

🎉

> .from starts growing on me, it's easy to read 👍

👊

> I'll thinker on the subclassing later, but I agree that it's a low prio

💯

THEtheChad commented:
Yikes! This PR is a lot. Feel free to take adequate time and ensure there aren't any mistakes (I already identified a bunch myself).

timdeschryver reviewed (approved):
I went over it and it looks good to me!
The naming/restructure makes a lot of sense now when it's complete 🤩

While we got the Fixture class now, do you think it would still make sense to have a `createFixture` method as the most basic utility?
I see the current fixture be created in e.g. a `beforeAll` or `beforeEach`, but maybe a `createFixture` still makes sense to use it directly within a test case?

Because it's already a big PR and I like everything about it, I'm already going to merge this in :)
